### PR TITLE
standard heartbeat backoff

### PR DESCRIPTION
**Pull request by emmett9001**
This pull request implements Parsely's standard heartbeat backoff mechanism in the Sampler.

### CODE REVIEW

File: ParselyTracker/Sampler.swift [hunk 1]
```diff
@@ -64,6 +65,7 @@ class Sampler {
                   key: key,
                   accumulatedTime: TimeInterval(0),
                   totalTime: TimeInterval(0),
+                  firstSampleTime: Date(),
```
Comment (aldraco) [inline ParselyTracker/Sampler.swift]: Nitpicky, perhaps -- I wonder if `firstSampleTime` and `lastSampleTime` sound too much like endpoints when put next to each other, and something like `mostRecentSample` (or something more brief) could help highlight that we aren't trying to define some interval of time where sampling takes place.

File: ParselyTracker/Sampler.swift [hunk 2]
```diff
@@ -124,6 +126,17 @@ class Sampler {
         }
         self.samplerTimer = Timer.scheduledTimer(withTimeInterval: SAMPLE_RATE, repeats: false) { timer in self.sample() }
     }
+    
+    private func backoff(existingTimeout: TimeInterval,
+                         totalTrackedTime: TimeInterval) -> TimeInterval
+    {
+        let offsetMatchingBaseInterval: TimeInterval = TimeInterval(35)
+        let backoffProportion: Double = 0.3
```
Comment (aldraco) [inline ParselyTracker/Sampler.swift]: This might be a good global/class value.

File: ParselyTracker/Sampler.swift [hunk 3]
```diff
@@ -124,6 +126,17 @@ class Sampler {
         }
         self.samplerTimer = Timer.scheduledTimer(withTimeInterval: SAMPLE_RATE, repeats: false) { timer in self.sample() }
     }
+    
+    private func backoff(existingTimeout: TimeInterval,
+                         totalTrackedTime: TimeInterval) -> TimeInterval
+    {
+        let offsetMatchingBaseInterval: TimeInterval = TimeInterval(35)
```
Comment (aldraco) [inline ParselyTracker/Sampler.swift]: Took me a second to remember what all of these numbers mean (35, 0.3).

Let's make those globals and/or put some comments/docstrings that explain where they came from and how they're used. 

I also think changing the name from `backoff()`, which implies this function does the action of assigning the new value as well as finding it, to something that limits the reader's expectation to "computes a value and returns it" would be helpful.

File: ParselyTracker/Sampler.swift [hunk 4]
```diff
@@ -133,7 +146,11 @@ class Sampler {
             heartbeatFn(data: trackedData, enableHeartbeats: true)
         }
         trackedData.accumulatedTime = 0
-        trackedData.heartbeatTimeout = TimeInterval(min(MAX_TIME_BETWEEN_HEARTBEATS, trackedData.heartbeatTimeout! * 1.25))
+        let currentTime: Date = Date();
```
Comment (aldraco) [inline ParselyTracker/Sampler.swift]: These two lines could probably be combined into one for readability.

File: ParselyTracker/Sampler.swift (hunk 3 above)
Comment (emmett9001) [reply ParselyTracker/Sampler.swift]: I don't think that making these global adds anything to how self-explanatory these variables are. `offsetMatchingBaseInterval` is my attempt at a variable name that clearly explains the magic of the number `35` in this context. The problem is that `35` and `.3` here are both numbers that resulted from a bunch of tinkering in a simulator and don't have clear meaning in themselves.

I agree that `backoff` could be named a lot better here.

File: ParselyTracker/Sampler.swift (hunk 1 above)
Comment (emmett9001) [reply ParselyTracker/Sampler.swift]: I can see better names for `lastSampleTime`, but changing that seems out of scope for this pull request. I've added it to https://github.com/Parsely/engineering/issues/3042

File: ParselyTracker/Sampler.swift [hunk 5]
```diff
@@ -124,6 +128,15 @@ class Sampler {
         }
         self.samplerTimer = Timer.scheduledTimer(withTimeInterval: SAMPLE_RATE, repeats: false) { timer in self.sample() }
     }
+    
+    private func getHeartbeatInterval(existingTimeout: TimeInterval,
+                                      totalTrackedTime: TimeInterval) -> TimeInterval
+    {
+        let totalWithOffset: TimeInterval = totalTrackedTime + offsetMatchingBaseInterval
```
Comment (aldraco) [inline ParselyTracker/Sampler.swift]: New name reads much more easily. A comment explaining the algorithm would still be nice, though having it encapsulated here with an explanatory name implies enough.

**Comment by emmett9001** [comment]
This is blocked by the equivalent change in the closed-source Parsely JS SDK.

**Pull request merged**